src/webgpu/api/operation/memory_sync/buffer/multiple_buffers.spec.ts

```diff
@@ -27,9 +42,56 @@ g.test('rw')
     Perform a 'read' operations on multiple buffers, followed by a 'write' operation.
     Operations are separated by a 'boundary' (pass, encoder, queue-op, etc.).
     Test that the results are synchronized.
-    The read should not see the contents written by the subsequent write.`
+    The read should not see the contents written by the subsequent write.
+  `
   )
-  .unimplemented();
+  .params(u =>
+    u //
+      .combine('boundary', kOperationBoundaries)
+      .expand('_context', p => kBoundaryInfo[p.boundary].contexts)
+      .expandWithParams(function* ({ _context }) {
+        for (const readOp of kAllReadOps) {
+          for (const writeOp of kAllWriteOps) {
+            if (checkOpsValidForContext([readOp, writeOp], _context)) {
+              yield {
+                readOp,
+                readContext: _context[0],
+                writeOp,
+                writeContext: _context[1],
+              };
+            }
+          }
+        }
+      })
+  )
+  .fn(async t => {
+    const { readContext, readOp, writeContext, writeOp, boundary } = t.params;
+    const helper = new OperationContextHelper(t);
+
+    const srcBuffers: GPUBuffer[] = [];
+    const dstBuffers: GPUBuffer[] = [];
+
+    const kSrcBufferCount = 4;
+    for (let i = 0; i < kSrcBufferCount; i++) {
+      const { srcBuffer, dstBuffer } = await t.createBuffersForReadOp(readOp, kSrcValue, kOpValue);
+
+      srcBuffers.push(srcBuffer);
+      dstBuffers.push(dstBuffer);
+
+      await t.createIntermediateBuffersAndTexturesForWriteOp(writeOp, 0, kOpValue);
+
+      // The read op will read from src buffer and write to dst buffer based on what it reads.
+      // The write op will write the given op value into src buffer as well.
+      // The write op happens after read op. So we are expecting the src value to be in the dst
+      // buffer.
+      t.encodeReadOp(helper, readOp, readContext, srcBuffers[i], dstBuffers[i]);
+      helper.ensureBoundary(boundary);
+      t.encodeWriteOp(helper, writeOp, writeContext, srcBuffers[0], 0, kOpValue);
+      helper.ensureSubmit();
```

Comment:
(r = read, w = write, c = check)
This will result in r,w,c,r,w,c,r,w,c,r,w,c. I think the intent of the test is to do r,r,r,r,w,w,w,w,c,c,c,c so that work is more interleaved. For example, an implementation might coalesce 4 barriers (one for each r-w pair) into 1 barrier to cover all 4, and we want to make sure that optimization is correct.

Reply:
Thanks for the comment. So I update this PR to do `r,r,r,r -> a boundary -> w,w,w,w -> c,c,c,c` to ensure a boundary can separate operations in multiple buffers. Does it make sense?

Reply:
Yes, lgtm!